📚 docs(granite-speech): add comprehensive usage examples

gorkachea opened the pull request:
# docs(granite-speech): add comprehensive usage examples

Resolves the TODO (@alex-jw-brooks) by adding complete usage documentation for Granite Speech model now that it's released and compatible with transformers.

Added examples for:
- Basic speech transcription
- Speech-to-text with additional context
- Batch processing multiple audio files
- Tips for best results (audio format, LoRA adapter, memory optimization)

This helps users get started with the Granite Speech multimodal model by providing practical, copy-paste-ready code examples for common use cases.

Replaces TODO comment on line 44 with ~100 lines of comprehensive documentation following the patterns used in other multimodal model docs.

## What does this PR do?

This PR resolves an existing TODO in the Granite Speech documentation by adding comprehensive usage examples now that the model is released and compatible with transformers.

**Problem**: The documentation contained a placeholder TODO comment asking for usage examples once the model was released. Since `ibm-granite/granite-3.2-8b-speech` is now available, users need practical examples to get started.

**Solution**: Added a complete "Usage example" section with three practical code examples:
1. Basic speech transcription - simple audio-to-text
2. Speech-to-text with additional context - using text prompts with audio
3. Batch processing - efficient handling of multiple audio files
4. Tips section - audio format, LoRA adapter info, optimization guidance

**Impact**: Users can now quickly get started with Granite Speech without searching external resources. The examples follow the same structure as other multimodal models (SAM2, SmolVLM, LLaVA) for consistency.

## Before submitting

- [x] This PR fixes a typo or improves the docs - **Yes, this is documentation enhancement**
- [x] Did you read the contributor guideline, Pull Request section? **Yes**
- [ ] Was this discussed/approved via a Github issue or the forum? **No - this directly resolves an existing TODO comment left by @alex-jw-brooks in the documentation**
- [x] Did you make sure to update the documentation with your changes? **Yes - this PR is entirely documentation**
- [ ] Did you write any new necessary tests? **N/A - documentation only change, no code changes**

## Who can review?

@alex-jw-brooks (original TODO author)
@stevhliu (documentation)
@zucchini-nlp (multimodal models)

---

**Additional Context:**

This PR adds ~105 lines of documentation to replace a single TODO comment. All code examples:
- Follow transformers conventions
- Are syntactically correct
- Match patterns used in other multimodal model docs
- Cover the most common use cases for this model

The Granite Speech model is a multimodal speech-to-text model from IBM that's now fully integrated into transformers, so users need these examples to get started effectively.

gorkachea committed:
📚 docs(granite-speech): add comprehensive usage examples

Resolves the TODO (@alex-jw-brooks) by adding complete usage documentation
for Granite Speech model now that it's released and compatible with transformers.

Added examples for:
- Basic speech transcription
- Speech-to-text with additional context
- Batch processing multiple audio files
- Tips for best results (audio format, LoRA adapter, memory optimization)

This helps users get started with the Granite Speech multimodal model
by providing practical, copy-paste-ready code examples for common use cases.

Replaces TODO comment on line 44 with ~100 lines of comprehensive
documentation following the patterns used in other multimodal model docs.

zucchini-nlp commented on docs/source/en/model_doc/granite_speech.md:
```diff
@@ -41,7 +41,111 @@ This model was contributed by [Alexander Brooks](https://huggingface.co/abrooks9
+
+# Prepare inputs with text prompt
+text_prompt = "Transcribe the following audio:"
+audio_input = "path/to/audio.wav"
```
prob we need to apply a chat template and format the prompt

zucchini-nlp reviewed (commented):
cc @eustlb for audio

gorkachea committed:
Address review feedback: add chat template usage and move model-specific tips

- Added proper chat template formatting in the second example (per @zucchini-nlp feedback)
- Removed generic LLM tips (temperature, batch size, memory)
- Moved Granite Speech-specific tips (audio format, LoRA adapter) to Usage tips section

This keeps the documentation focused on model-specific features rather than general LLM knowledge.

gorkachea committed:
Merge branch 'main' into docs/add-granite-speech-usage-example

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_42125). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

gorkachea committed:
Merge branch 'main' into docs/add-granite-speech-usage-example

zucchini-nlp reviewed (approved):
lgtm!

eustlb commented on docs/source/en/model_doc/granite_speech.md:
```diff
@@ -40,8 +40,108 @@ This model was contributed by [Alexander Brooks](https://huggingface.co/abrooks9
+    text=text,
+    audio="path/to/audio.wav",
+    return_tensors="pt"
+).to(model.device)
```
why can't we do apply_chat_template with tokenize=True directly? That would be the standard way of doing it

gorkachea replied:
Good point! I checked Qwen2Audio (similar audio+text model) and they use tokenize=False → processor(text, audio) to let the processor coordinate tokenization with audio features. The current code follows this pattern, but let me know if you'd prefer a different approach for Granite Speech - always open to learning the best practice here!

eustlb replied:
I see 🤗 Indeed, the Qwen2Audio snippets are out-of-date (if I recall audio loading was not directly possible from the chat template at that time). You should be able to do something like:
```python
conversation = [
    {
        "role": "user",
        "content": [
            {
                "type": "audio",
                "url": "https://huggingface.co/datasets/eustlb/audio-samples/resolve/main/dude_where_is_my_car.wav",
            },
            {"type": "text", "text": "What can you tell me about this audio?"},
        ],
    }
]

inputs = processor.apply_chat_template(conversation, tokenize=True)
```

gorkachea replied:
Thanks for clarifying! 🙏 You're absolutely right - I've updated the example to use the modern pattern with `processor.apply_chat_template(conversation, tokenize=True)` directly. The audio is now embedded in the conversation structure with `{"type": "audio", "audio": audio}` and the processor handles everything in a single call.

The example now follows the same pattern as your suggestion and should be up-to-date with current best practices. Thanks for the guidance on this!

gorkachea committed:
docs: use datasets library for working audio examples

Address review feedback by replacing placeholder audio paths with real
examples using hf-internal-testing/librispeech_asr_dummy dataset. This
makes all code examples copy-paste ready and reproducible.

- Add datasets import to all three examples
- Replace 'path/to/audio.wav' with actual dataset loading
- Ensure proper audio sampling rate handling

Co-authored-by: eustlb <eustache.leblond@gmail.com>

gorkachea committed:
📚 docs: use modern chat template pattern with tokenize=True for audio

eustlb reviewed (approved):
LGTM, thanks for iterating!
